packages/ra-core/src/form/FormDataConsumer.js

```diff
@@ -3,6 +3,7 @@ import { connect } from 'react-redux';
 import { getFormValues } from 'redux-form';
 
 const REDUX_FORM_NAME = 'record-form';
+const ARRAY_SOURCE_REGEX = /^(.*)\[(\d*)\]/;
```

Comment:
I think the actual shape differs a bit. When I look at `SimpleFormIterator`'s way of recreating the source, it reads like this:

```
${member}.${input.props.source}
```

Where `member` is something like `backlinks[3]`.

So when used with a `SimpleFormIterator`, which has no `source`, the result has a closing dot (``backlinks[3].`). This `source` doesn't make sense, and I don't think you should use it inside `SimpleFormIterator`. Instead, you should detect in `SimpleFormIterator` when the child has no `source`, and inject the `index` instead.

Reply:
In the case of the `FormDataConsumer`, which has no `source`, we end up with `backlinks[3].undefined`

Reply:
Thanks to your inputs, I ended up not using any regex at all :+1: